File: pyclient/confluo/rpc/rpc_client.py
import logging
import rpc_type_conversions
import rpc_configuration_params
import rpc_service
from rpc_record_batch_builder import rpc_record_batch_builder
from rpc_stream import record_stream, alert_stream

from ttypes import *
from thrift.transport import TTransport, TSocket
from thrift.protocol.TBinaryProtocol import TBinaryProtocol

class rpc_client:
    """
    Client for Confluo through RPC
    """
    def __init__(self, host='localhost', port=9090):
        """
        Initializes the rpc client to the specified host and port
        Args:
            host: The host for the client
            port: The port number to communicate through
        """
        logging.basicConfig(level=logging.INFO)  # TODO: Read from configuration file
        self.LOG = logging.getLogger(__name__)
        self.connect(host, port)
        self.cur_multilog_id_ = -1

    def close(self):
        """
        Closes the rpc client
        """
        self.disconnect()

    def connect(self, host, port):
        """
        Connects the rpc client to the specified host and port
        Args:
            host: The host of the client
            port: The port number to communicate through
        """
        self.LOG.info("Connecting to %s:%d", host, port)
        self.socket_ = TSocket.TSocket(host, port)
        self.transport_ = TTransport.TBufferedTransport(self.socket_)
        self.protocol_ = TBinaryProtocol(self.transport_)
        self.client_ = rpc_service.Client(self.protocol_)
        self.transport_.open()
        self.client_.register_handler()

    def disconnect(self):
        """
        Disconnects the rpc client from the host and port
        """
        if self.transport_.isOpen():
            host = self.socket_.host
            port = self.socket_.port
            self.LOG.info("Disconnecting from %s:%d", host, port)
            self.client_.deregister_handler()
            self.transport_.close()

    def create_atomic_multilog(self, atomic_multilog_name, schema, storage_mode):
        """
        Creates an atomic multilog for this client
        Args:
            name: The name of the atomic multilog to create
            schema: The schema for the atomic multilog
            storage_mode: The mode for storage
        """
        self.cur_schema_ = schema
        rpc_schema = rpc_type_conversions.convert_to_rpc_schema(schema)
        self.cur_multilog_id_ = self.client_.create_atomic_multilog(atomic_multilog_name, rpc_schema, storage_mode)
        
    def set_current_atomic_multilog(self, atomic_multilog_name):
        """
        Sets the atomic multilog to the desired atomic multilog
        Args:
            atomic_multilog_name: The name of atomic multilog to set
            the current atomic multilog to
        """
        info = self.client_.get_atomic_multilog_info(atomic_multilog_name) 
        self.cur_schema_ = rpc_type_conversions.convert_to_schema(info.schema)
        self.cur_multilog_id_ = info.atomic_multilog_id
        
    def remove_atomic_multilog(self):
        """
        Removes an atomic multilog from the client
        Raises:
            ValueError
        """
        if self.cur_multilog_id_ == -1:
            raise ValueError("Must set atomic multilog first.")
        self.client_.remove_atomic_multilog(self.cur_multilog_id_)
        self.cur_multilog_id_ = -1

    def run_command(self, json_command):
        """
        Executes a command specified by a JSON string
        Raises:
            ValueError
        """
        if self.cur_multilog_id_ == -1:
            raise ValueError("Must set atomic multilog first.")
        self.client_.run_command(self.cur_multilog_id_, json_command)
    
    def add_index(self, field_name, bucket_size=1):
        """
        Adds an index to the atomic multilog
        Raises:
            ValueError
        """
        if self.cur_multilog_id_ == -1:
            raise ValueError("Must set atomic multilog first.")
        self.client_.add_index(self.cur_multilog_id_, field_name, bucket_size)

    def remove_index(self, field_name):
        """
        Removes an index from the atomic multilog
        Args:
            field_name: The name of the associated field
        Raises:
            ValueError
        """
        if self.cur_multilog_id_ == -1:
            raise ValueError("Must set atomic multilog first.")
        self.client_.remove_index(self.cur_multilog_id_, field_name)

    def add_filter(self, filter_name, filter_expr):
        """
        Adds a filter to the atomic multilog
        Args:
            filter_name: The name of the filter
            filter_expr: The filter expression
        Raises:
            ValueError
        """
        if self.cur_multilog_id_ == -1:
            raise ValueError("Must set atomic multilog first.")
        self.client_.add_filter(self.cur_multilog_id_, filter_name, filter_expr)

    def remove_filter(self, filter_name):
        """
        Removes a filter from the atomic multilog
        Args:
            filter_name: The name of the filter
        Raises:
            ValueError
        """
        if self.cur_multilog_id_ == -1:
            raise ValueError("Must set atomic multilog first.")
        self.client_.remove_filter(self.cur_multilog_id_, filter_name)
        
    def add_aggregate(self, aggregate_name, filter_name, aggregate_expr):
        """
        Adds an aggregate to the atomic multilog
        Args:
            aggregate_name: The name of the aggregate
            filter_name: The name of the filter
            aggregate_expr: The aggregate expression
        Raises:
            ValueError
        """
        if self.cur_multilog_id_ == -1:
            raise ValueError("Must set atomic multilog first.")
        self.client_.add_aggregate(self.cur_multilog_id_, aggregate_name, filter_name, aggregate_expr)

    def remove_aggregate(self, aggregate_name):
        """
        Removes an aggregate from the atomic multilog
        Args:
            aggregate_name: The name of the aggregate
        Raises:
            ValueError
        """
        if self.cur_multilog_id_ == -1:
            raise ValueError("Must set atomic multilog first.")
        self.client_.remove_aggregate(self.cur_multilog_id_, aggregate_name)

    def install_trigger(self, trigger_name, trigger_expr):
        """
        Adds a trigger to the atomic multilog
        Args:
            trigger_name: The name of the trigger to add
            trigger_expr: The trigger expression
        Raises:
            ValueError
        """
        if self.cur_multilog_id_ == -1:
            raise ValueError("Must set atomic multilog first.")
        self.client_.add_trigger(self.cur_multilog_id_, trigger_name, trigger_expr)

    def remove_trigger(self, trigger_name):
        """
        Removes a trigger from the atomic multilog
        Args:
            trigger_name: The name of the trigger
        Raises:
            ValueError
        """
        if self.cur_multilog_id_ == -1:
            raise ValueError("Must set atomic multilog first.")
        self.client_.remove_trigger(self.cur_multilog_id_, trigger_name)
        
    def get_batch_builder(self):
        """
        Gets a new record batch builder
        """
        return rpc_record_batch_builder()
        
    def write(self, record):
        """
        Writes a record to the atomic multilog
        Args:
            record: The record to write
        Raises:
            ValueError
        """
        if self.cur_multilog_id_ == -1:
            raise ValueError("Must set atomic multilog first.")
        if len(record) != self.cur_schema_.record_size_:
            raise ValueError("Record must be of length " + str(self.cur_schema_.record_size_))
        self.client_.append(self.cur_multilog_id_, record)
            
    def read(self, offset):
        """
        Reads data from a specified offset
        Args:
            offset: The offset from the log to read from
        Raises:
            ValueError
        Returns:
            The data at the offset
        """
        if self.cur_multilog_id_ == -1:
            raise ValueError("Must set atomic multilog first.")
        return self.client_.read(self.cur_multilog_id_, offset, self.cur_schema_.record_size_)
    
    def get_aggregate(self, aggregate_name, begin_ms, end_ms):
        """
        Gets an aggregate from the atomic multilog
        Args:
            aggregate_name: The name of the aggregate
            begin_ms: The beginning time in milliseconds
            end_ms: The end time in milliseconds
        Raises:
            ValueError
        Returns:
            The aggregate
        """
        if self.cur_multilog_id_ == -1:
            raise ValueError("Must set atomic multilog first.")
        return self.client_.query_aggregate(self.cur_multilog_id_, aggregate_name, begin_ms, end_ms)

    def execute_filter(self, filter_expr):
        """
        Executes a specified filter
        Args:
            filter_expr: The filter expression
        Raises:
            ValueError
        Returns:
            Record stream containing the data
        """
        if self.cur_multilog_id_ == -1:
            raise ValueError("Must set atomic multilog first.")
        handle = self.client_.adhoc_filter(self.cur_multilog_id_, filter_expr)
        return record_stream(self.cur_multilog_id_, self.cur_schema_, self.client_, handle)

    def query_filter(self, filter_name, begin_ms, end_ms, filter_expr = ""):
        """
        Queries a filter
        Args:
            filter_name: The name of the filter
            begin_ms: The beginning time in milliseconds
            end_ms: The end time in milliseconds
            filter_expr: The filter expression
        Raises:
            ValueError
        Returns:
            A record stream containing the results of the filter
        """
        if self.cur_multilog_id_ == -1:
            raise ValueError("Must set atomic multilog first.")
        if filter_expr == "":
            handle = self.client_.predef_filter(self.cur_multilog_id_, filter_name, begin_ms, end_ms)
            return record_stream(self.cur_multilog_id_, self.cur_schema_, self.client_, handle)
        else:
            handle = self.client_.combined_filter(self.cur_multilog_id_, filter_name, filter_expr, begin_ms, end_ms)
            return record_stream(self.cur_multilog_id_, self.cur_schema_, self.client_, handle)

    def get_alerts(self, begin_ms, end_ms, trigger_name = ""):
        """
        Gets the alerts
        Args:
            begin_ms: The beginning time in milliseconds
            end_ms: The end time in milliseconds
            trigger_name: The name of the trigger
        Raises:
            ValueError
        Returns:
            A stream of alerts
        """
        if self.cur_multilog_id_ == -1:
            raise ValueError("Must set atomic multilog first.")
        if trigger_name == "":
            handle = self.client.alerts_by_time(self.cur_multilog_id_, handle, begin_ms, end_ms)
            return alert_stream(self.cur_multilog_id_, self.cur_schema_, self.client_, handle)
        else:
            handle = self.client.alerts_by_time(self.cur_multilog_id_, handle, trigger_name, begin_ms, end_ms)
            return alert_stream(self.cur_multilog_id_, self.cur_schema_, self.client_, handle)

    def num_records(self):
        """
        Gets the number of records in the atomic multilog
        Raises:
            ValueError
        Returns:
            The number of records
        """
        if self.cur_multilog_id_ == -1:
            raise ValueError("Must set atomic multilog first.")
        return self.client_.num_records(self.cur_multilog_id_)

File: javaclient/src/main/java/confluo/rpc/RPCClient.java
package confluo.rpc;

import org.apache.thrift.TException;
import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.transport.TSocket;
import org.apache.thrift.transport.TTransport;

import java.nio.ByteBuffer;
import java.util.List;

/**
 * Client for Confluo through RPC
 */
public class RPCClient {

  private TTransport transport;
  private rpc_service.Client client;
  private long curMultilogId;
  private rpc_atomic_multilog_info info;
  private Schema curSchema;

  /**
   * Initializes the rpc client to the specified host and port
   *
   * @param host The host for the client
   * @param port The port number to communicate through
   * @throws TException Cannot connect to host
   */
  public RPCClient(String host, int port) throws TException {
    connect(host, port);
    curMultilogId = -1;
  }

  /**
   * Closes the rpc client
   *
   * @throws TException Cannot disconnect
   */
  public void close() throws TException {
    disconnect();
  }

  /**
   * Connects the rpc client to the specified host and port
   *
   * @param host The host of the client
   * @param port The port number to communicate through
   * @throws TException Cannot connect
   */
  public void connect(String host, int port) throws TException {
    transport = new TSocket(host, port);
    TBinaryProtocol protocol = new TBinaryProtocol(transport);
    client = new rpc_service.Client(protocol);
    transport.open();
    client.registerHandler();
  }

  /**
   * Disconnects the rpc client from the host and port
   *
   * @throws TException Cannot disconnect
   */
  public void disconnect() throws TException {
    if (transport.isOpen()) {
      client.deregisterHandler();
      transport.close();
    }
  }

  /**
   * Creates an atomic multilog for this client
   *
   * @param atomicMultilogName The name of the atomic multilog to create
   * @param schema             The schema for the atomic multilog
   * @param storageMode        The mode for storage
   * @throws TException Cannot create the atomic multilog
   */
  public void createAtomicMultilog(String atomicMultilogName, Schema schema, rpc_storage_mode storageMode) throws TException {
    this.curSchema = schema;
    List<rpc_column> rpcSchema = RPCTypeConversions.convertToRPCSchema(schema);
    curMultilogId = client.createAtomicMultilog(atomicMultilogName, rpcSchema, storageMode);
  }

  /**
   * Sets the atomic multilog to the desired atomic multilog
   *
   * @param atomicMultilogName The name of atomic multilog to set the current atomic multilog to
   * @throws TException Cannot set the atomic multilog
   */
  public void setCurrentAtomicMultilog(String atomicMultilogName) throws TException {
    info = client.getAtomicMultilogInfo(atomicMultilogName);
    curSchema = RPCTypeConversions.convertToSchema(info.getSchema());
    curMultilogId = info.getId();
  }

  /**
   * Removes an atomic multilog from the client
   *
   * @throws TException Cannot remove the atomic multilog
   */
  public void removeAtomicMultilog() throws TException {
    if (curMultilogId == -1) {
      throw new IllegalStateException("Must set Atomic Multilog first");
    }
    client.removeAtomicMultilog(curMultilogId);
    curMultilogId = -1;
  }

  /**
   * Adds an index to the atomic multilog
   *
   * @param fieldName  The name of the associated field
   * @param bucketSize The bucket in the multilog to add the index
   * @throws TException Cannot add the index
   */
  public void addIndex(String fieldName, double bucketSize) throws TException {
    if (curMultilogId == -1) {
      throw new IllegalStateException("Must set Atomic Multilog first");
    }
    client.addIndex(curMultilogId, fieldName, bucketSize);
  }

  /**
   * Removes an index from the atomic multilog
   *
   * @param fieldName The name of the associated field
   * @throws TException Cannot remove the index
   */
  public void removeIndex(String fieldName) throws TException {
    if (curMultilogId == -1) {
      throw new IllegalStateException("Must set Atomic Multilog first");
    }
    client.removeIndex(curMultilogId, fieldName);
  }

  /**
   * Adds a filter to the atomic multilog
   *
   * @param filterName The name of the filter
   * @param filterExpr The filter expression
   * @throws TException Cannot add filter
   */
  public void addFilter(String filterName, String filterExpr) throws TException {
    if (curMultilogId == -1) {
      throw new IllegalStateException("Must set Atomic Multilog first");
    }
    client.addFilter(curMultilogId, filterName, filterExpr);
  }

  /**
   * Removes a filter from the atomic multilog
   *
   * @param filterName The name of the filter
   * @throws TException Cannot remove the filter
   */
  public void removeFilter(String filterName) throws TException {
    if (curMultilogId == -1) {
      throw new IllegalStateException("Must set Atomic Multilog first");
    }
    client.removeFilter(curMultilogId, filterName);
  }

  /**
   * Adds an aggregate to the atomic multilog
   *
   * @param aggregateName The name of the aggregate
   * @param filterName    The name of the filter
   * @param aggregateExpr The aggregate expression
   * @throws TException Cannot add the aggregate
   */
  public void addAggregate(String aggregateName, String filterName, String aggregateExpr) throws TException {
    if (curMultilogId == -1) {
      throw new IllegalStateException("Must set Atomic Multilog first");
    }
    client.addAggregate(curMultilogId, aggregateName, filterName, aggregateExpr);
  }

  /**
   * Removes an aggregate from the atomic multilog
   *
   * @param aggregateName The name of the aggregate
   * @throws TException Cannot remove the aggregate
   */
  public void removeAggregate(String aggregateName) throws TException {
    if (curMultilogId == -1) {
      throw new IllegalStateException("Must set Atomic Multilog first");
    }
    client.removeAggregate(curMultilogId, aggregateName);
  }

  /**
   * Adds a trigger to the atomic multilog
   *
   * @param triggerName The name of the trigger to add
   * @param triggerExpr The trigger expression
   * @throws TException Cannot add the trigger
   */
  public void installTrigger(String triggerName, String triggerExpr) throws TException {
    if (curMultilogId == -1) {
      throw new IllegalStateException("Must set Atomic Multilog first");
    }
    client.addTrigger(curMultilogId, triggerName, triggerExpr);
  }

  /**
   * Removes a trigger from the atomic multilog
   *
   * @param triggerName The name of the trigger
   * @throws TException Cannot remove the trigger
   */
  public void removeTrigger(String triggerName) throws TException {
    if (curMultilogId == -1) {
      throw new IllegalStateException("Must set Atomic Multilog first");
    }
    client.removeTrigger(curMultilogId, triggerName);
  }

  /**
   * Gets a new record batch builder
   *
   * @return The RPC record batch builder
   */
  public RPCRecordBatchBuilder getBatchBuilder() {
    return new RPCRecordBatchBuilder(curSchema);
  }

  /**
   * Writes a record to the atomic multilog
   *
   * @param record The record to write
   * @throws TException Cannot append the record
   */
  public void append(ByteBuffer record) throws TException {
    if (curMultilogId == -1) {
      throw new IllegalStateException("Must set Atomic Multilog first");
    }
    if (record.array().length != curSchema.getRecordSize()) {
      throw new IllegalStateException("Record size incorrect; expected=" + curSchema.getRecordSize()
          + ", got=" + record.array().length);
    }
    client.append(curMultilogId, record);
  }

  /**
   * Reads data from a specified offset
   *
   * @param offset The offset from the log to read from
   * @return The data at the offset
   * @throws TException Cannot read the record
   */
  public ByteBuffer read(long offset) throws TException {
    if (curMultilogId == -1) {
      throw new IllegalStateException("Must set Atomic Multilog first");
    }
    return client.read(curMultilogId, offset, 1);
  }

  /**
   * Gets an aggregate from the atomic multilog
   *
   * @param aggregateName The name of the aggregate
   * @param beginMs       The beginning time in milliseconds
   * @param endMs         The end time in milliseconds
   * @return The aggregate
   * @throws TException Cannot get the aggregate
   */
  public String getAggregate(String aggregateName, long beginMs, long endMs) throws TException {
    if (curMultilogId == -1) {
      throw new IllegalStateException("Must set Atomic Multilog first");
    }
    return client.queryAggregate(curMultilogId, aggregateName, beginMs, endMs);
  }

  /**
   * Executes a specified filter
   *
   * @param filterExpr The filter expression
   * @return Record stream containing the data
   * @throws TException Cannot execute the filter
   */
  public RecordStream executeFilter(String filterExpr) throws TException {
    if (curMultilogId == -1) {
      throw new IllegalStateException("Must set Atomic Multilog first");
    }
    rpc_iterator_handle handle = client.adhocFilter(curMultilogId, filterExpr);
    return new RecordStream(curMultilogId, curSchema, client, handle);
  }

  /**
   * Queries a filter
   *
   * @param filterName The name of the filter
   * @param beginMs    The beginning time in milliseconds
   * @param endMs      The end time in milliseconds
   * @param filterExpr The filter expression
   * @return A record stream containing the results of the filter
   * @throws TException Cannot query the filter
   */
  public RecordStream queryFilter(String filterName, long beginMs, long endMs, String filterExpr) throws TException {
    if (curMultilogId == -1) {
      throw new IllegalStateException("Must set Atomic Multilog first");
    }
    if (filterExpr.equals("")) {
      rpc_iterator_handle handle = client.predefFilter(curMultilogId, filterName, beginMs, endMs);
      return new RecordStream(curMultilogId, curSchema, client, handle);
    } else {
      rpc_iterator_handle handle = client.combinedFilter(curMultilogId, filterName, filterExpr, beginMs, endMs);
      return new RecordStream(curMultilogId, curSchema, client, handle);
    }
  }

  /**
   * Gets the alerts
   *
   * @param beginMs     The beginning time in milliseconds
   * @param endMs       The end time in milliseconds
   * @param triggerName The name of the trigger
   * @return A stream of alerts
   * @throws TException Cannot get the alerts
   */
  public AlertStream getAlerts(long beginMs, long endMs, String triggerName) throws TException {
    if (curMultilogId == -1) {
      throw new IllegalStateException("Must set Atomic Multilog first");
    }
    if (triggerName.equals("")) {
      rpc_iterator_handle handle = client.alertsByTime(curMultilogId, beginMs, endMs);
      return new AlertStream(curMultilogId, client, handle);
    } else {
      rpc_iterator_handle handle = client.alertsByTime(curMultilogId, beginMs, endMs);
      return new AlertStream(curMultilogId, client, handle);
    }
  }

  /**
   * Gets the number of records in the atomic multilog
   *
   * @return The number of records
   * @throws TException Cannot get the number of records
   */
  public long numRecords() throws TException {
    if (curMultilogId == -1) {
      throw new IllegalStateException("Must set Atomic Multilog first");
    }
    return client.numRecords(curMultilogId);
  }
}
